src/rez/developer_package.py

```diff
@@ -152,6 +164,34 @@ def get_reevaluated(self, objects):
         with set_objects(objects):
             return self.from_path(self.root)
 
+    def re_evaluate_variant(self, variant, build_context):
+        """Re-evaluate variant with resolved build context
+
+        This doesn't return new variant, but swapping the given variant's
+        resource in-place with re-created one.
+
+        Args:
+            variant (`Variant`): A variant of this package
+            build_context (`ResolvedContext`): build resolved context
+
+        Returns:
+            None
+        """
+        data = self.validated_data()
+
+        data = evaluate_directive_requires(data,
+                                           self.directives,
+                                           build_context)
+        # re-evaluate variant
+        #
+        package_cls = type(self)
+        re_evaluated_package = create_package(self.name, data, package_cls)
+        re_evaluated_variant = re_evaluated_package.get_variant(variant.index)
+        # swapping resource
+        variant._parent = re_evaluated_package
+        variant.context = re_evaluated_package.context
+        variant.wrapped = re_evaluated_variant.wrapped
+
```

Comment:
About this *variant resource swapping*, it is here because, the `Variant` object doesn't really carry the variated data by itself, but asking for it's parent package that has been evaluated for that variant. This behavior is okay for installed package, but not for developer package in build-time.

Maybe we should re-implement `DeveloperPackage.iter_variants` ? That could spawn variant with re-evaluated package as it's parent. With that, I think the code above could change to just retrieve the variant's parent package and update it's data.

Reply:
Sorry, scratch that.

Got some time to revisit this again with a bit more understanding about how underlaying works and re-try to avoid current resource swapping hack for applying directives.

I have re-tried with the `variant.install(overrides=data)`, but the overrides is applied after checking existing package data, so the directives cannot be applied if previous installed exists.

https://github.com/nerdvegas/rez/blob/3f66e9ca1a0db227af608757e8dced08e88bd461/src/rezplugins/package_repository/filesystem.py#L1366-L1369

Then, I re-tried another route, updating `variant.parent` (aka package) data.

In `variant.install`, the package data is retrieved by `variant.parent.validated_data`, which does this
https://github.com/nerdvegas/rez/blob/3f66e9ca1a0db227af608757e8dced08e88bd461/src/rez/utils/data_utils.py#L627-L637

Based on schema key, the data was firstly collected from the object attributes, then from `_data`, but only if the key is not yet been collected.

So the directives must be applied back as object attributes :
```python
        # self.resource._data.update(data)  # won't work
        for k, v in data.items():
            setattr(self.resource, k, v)
```

After knowing these, using `overrides` to apply directives is not a good option, but using `setattr` is still not clear enough.

Re-implement `DeveloperPackage.validated_data` seems better, in there, we could clearly define which data in `self.resource._data` can take precedence, for example:
```python
def validated_data():
    data = super(DeveloperPackage, self).validated_data()
    data.update(data.pop("buildtime_updates", {}))
    return data
```